src/realm/object-store/shared_realm.cpp

```diff
@@ -1420,3 +1423,220 @@ AuditInterface* Realm::audit_context() const noexcept
 {
     return m_coordinator ? m_coordinator->audit_context() : nullptr;
 }
+
+namespace {
+
+/*********************************** PropId **********************************/
+
+struct PropId {
+    PropId(TableKey tk, ColKey ck, const Property* prop, const ObjectSchema* os, bool b)
+        : table_key(tk)
+        , col_key(ck)
+        , origin_prop(prop)
+        , target_schema(os)
+        , mandatory(b)
+    {
+    }
+    void expand(KeyPath& key_path, KeyPathArray& key_path_array) const;
+
+    TableKey table_key;
+    ColKey col_key;
+    const Property* origin_prop;
+    const ObjectSchema* target_schema;
+    std::vector<PropId> children;
+    bool mandatory;
+};
+
+
+void PropId::expand(KeyPath& key_path, KeyPathArray& key_path_array) const
+{
+    key_path.emplace_back(table_key, col_key);
+    if (children.empty()) {
+        key_path_array.push_back(key_path);
+    }
+    else {
+        for (auto& child : children) {
+            child.expand(key_path, key_path_array);
+        }
+    }
+    key_path.pop_back();
+}
+
+/****************************** KeyPathResolver ******************************/
+
+class KeyPathResolver {
+public:
+    KeyPathResolver(Group& g, const Schema& schema)
+        : m_group(g)
+        , m_schema(schema)
+    {
+    }
+
+    void resolve(const ObjectSchema* object_schema, const char* path)
```

Comment:
This could perhaps use some documentation to describe the difference between `resolve` and `expand`?